### ISSUE TITLE
[FEATURE] benchmark, this code runs more than 3x slower compare to pytorch

### ISSUE THREAD

**Issue by thegodone** [ml-explore/mlx#1531]
**Describe the bug**
MLX is much slower than Pytorch  & Tensorflow using GPU : (~ 0.46 sec per epoch for Pytorch & Tensorflow  versus 1.55 sec for MLX)

**To Reproduce**

follow instruction and codes from https://github.com/thegodone/apple_ai_model

**Expected behavior**
MLX is really slower than tensorflow and pytorch. 

**Desktop:**
 - OS: [MacOS 15.1]
 - mlx [0.19.1]
 - pytorch [2.5]
 - tensorflow [2.15]

**Comment by awni** [comment]
TLDR: with a few fixes + mx.compile it comes down to `Time: 0.3940401077270508` compared to PyTorch `Time: 0.7042069435119629` on an M1 Max.

More comments on this:

- Our dropout sampling was pretty slow. https://github.com/ml-explore/mlx/pull/1535 should improve the forward time a lot.

- Avoid constructs like:
```python
atom_neighbor = [atom_list[i][atom_degree_list[i]] for i in range(batch_size)]
atom_neighbor = mx.stack(atom_neighbor, axis=0)
 ``` 
which is completely un-vectorized and will be very slow.

You can do this in a single indexing op like so:
```python
atom_neighbor = mx.take_along_axis(atom_list[..., None, :], atom_degree_list[..., None], axis=-3)
```

- Put the model in `eval` mode when testing to avoid using dropout (`model.eval()`) and then back to training mode for training (`model.train()`).


With the dropout fix + switching those unvectorized ops, the training time goes down to: `Time: 0.47445106506347656`

Using `mx.compile` gets it down to `Time: 0.3940401077270508`

Here's roughly what that looks like:

```python
lr_schedule = cosineannealingwarmrestartfactor(initial_lr, restarts, decay_step, warmup_factor)

optimizer = optim.AdamW(learning_rate=lr_schedule, weight_decay=1**-weight_decay)

def loss_fn(y_hat, y):
    y = mx.reshape(y, y_hat.shape)
    return mx.mean(nn.losses.mse_loss(y_hat, y))

from functools import partial

state = [model, optimizer.state, mx.random.state]

def forward_fn(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, labels):
    _, y_hat = model(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask)
    loss = loss_fn(y_hat, labels)
    return loss, y_hat

@partial(mx.compile, inputs=state, outputs=state)
def step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, labels):
    loss_and_grad_fn = nn.value_and_grad(model, forward_fn)
    (loss, y_hat), grads = loss_and_grad_fn(
        x_atom=x_atom,
        x_bonds=x_bonds,
        x_atom_index=x_atom_index,
        x_bond_index=x_bond_index,
        x_mask=x_mask,
        labels=labels,
    )
    optimizer.update(model, grads)
    return loss
```

And in the train loop:
```python
    for counter, train_batch in enumerate(batch_list):
        batch_df = dataset.loc[train_batch,:]
        smiles_list = batch_df.cano_smiles.values
        y_val = mx.array(batch_df[tasks[0]].values)
        x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, smiles_to_rdkit_list = get_smiles_array(smiles_list,feature_dicts)
        x_atom = mx.array(x_atom)
        x_bonds = mx.array(x_bonds)
        x_atom_index = mx.array(x_atom_index)
        x_bond_index = mx.array(x_bond_index)
        x_mask = mx.array(x_mask)


        loss = step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, y_val)
        mx.eval(state)

**Comment by thegodone** [comment]
This is great thanks for the tips and improvement. the gather function aka take_along_axis was a mystery to set up, for me, this is great news.

**Comment by thegodone** [comment]
Do you think the GRUCell can even speed up or not ?

**Comment by awni** [comment]
>  Do you think the GRUCell can even speed up or not ?

Possibly... but I'm not sure it's the bottleneck in you model right now so I wouldn't over index on that. The best thing to do is figure out where the actual bottleneck is (sometimes this is not so easy) and focus on it.

**Comment by thegodone** [comment]
I try your modifications but I have an error during the compile I guess: 

````
device = mx.gpu # or mx.cpu
mx.set_default_device(device)

x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, smiles_to_rdkit_list = get_smiles_array([canonical_smiles_list[0]],feature_dicts)
num_atom_features = x_atom.shape[-1]
num_bond_features = x_bonds.shape[-1]
x_atom = mx.array(x_atom)
x_bonds = mx.array(x_bonds)
x_atom_index = mx.array(x_atom_index)
x_bond_index = mx.array(x_bond_index)
x_mask = mx.array(x_mask)

model = AttFP(radius, T, num_atom_features, num_bond_features,
            fingerprint_dim, output_units_num, p_dropout)



from functools import partial

def loss_fn(y_hat, y):
    y = mx.reshape(y, y_hat.shape)
    return mx.mean(nn.losses.mse_loss(y_hat, y))


state = [model, optimizer.state, mx.random.state]

def forward_fn(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, labels):
    _, y_hat = model(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask)
    loss = loss_fn(y_hat, labels)
    return loss, y_hat

@partial(mx.compile, inputs=state, outputs=state)
def step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, labels):
    loss_and_grad_fn = nn.value_and_grad(model, forward_fn)
    (loss, y_hat), grads = loss_and_grad_fn(
        x_atom=x_atom,
        x_bonds=x_bonds,
        x_atom_index=x_atom_index,
        x_bond_index=x_bond_index,
        x_mask=x_mask,
        labels=labels,
    )
    optimizer.update(model, grads)
    return loss
        
def train(dataset,e, batch_size=64, doprofile=False):
    print('train')
    loss_sum = 0.0
    np.random.seed(e)
    valList = np.arange(0,dataset.shape[0])
    #shuffle them
    np.random.shuffle(valList)
    batch_list = []
    for i in range(0, dataset.shape[0], batch_size):
        batch = valList[i:i+batch_size]
        batch_list.append(batch)   
    #print('iter per batch:',len(batch_list))
    if doprofile:
        pr = cProfile.Profile()
        pr.enable()  # Start profiling

    for counter, train_batch in enumerate(batch_list):
        batch_df = dataset.loc[train_batch,:]
        smiles_list = batch_df.cano_smiles.values
        y_val = mx.array(batch_df[tasks[0]].values)
        x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, smiles_to_rdkit_list = get_smiles_array(smiles_list,feature_dicts)
        x_atom = mx.array(x_atom)
        x_bonds = mx.array(x_bonds)
        x_atom_index = mx.array(x_atom_index)
        x_bond_index = mx.array(x_bond_index)
        x_mask = mx.array(x_mask)

        loss = step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, y_val)
        mx.eval(state)
        
    if doprofile:
        pr.disable()
        # Print profiling results
        s = io.StringIO()
        ps = pstats.Stats(pr, stream=s).sort_stats(pstats.SortKey.CUMULATIVE)
        ps.print_stats(200)  # Print top 10 results
        print(s.getvalue())    
    return loss_sum / len(dataset)

def test(test_dataset, batch_size=64):
    mse= 0.0
    valList = np.arange(0,test_dataset.shape[0])
    batch_list = []
    for i in range(0, test_dataset.shape[0], batch_size):
        batch = valList[i:i+batch_size]
        batch_list.append(batch) 
    
    model.eval()
        
    for counter, test_batch in enumerate(batch_list):
        batch_df = test_dataset.loc[test_batch,:]
        smiles_list = batch_df.cano_smiles.values
        
        x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, _ = get_smiles_array(smiles_list,feature_dicts)
        x_atom = mx.array(x_atom)
        x_bonds = mx.array(x_bonds)
        x_atom_index = mx.array(x_atom_index)
        x_bond_index = mx.array(x_bond_index)
        x_mask = mx.array(x_mask)
        y_val = mx.array(batch_df[tasks[0]].values)
        _, y_hat = model(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask)
        y_val = mx.reshape(y_val, y_hat.shape)
        mse += mx.square(y_hat - y_val).sum().item()
        
    val =  mse / len(test_dataset)
    model.train()
    return val, np.sqrt(val)


def epoch(e, batch_size=64):
    loss = train(train_df,e, batch_size=batch_size)
    train_mse, train_rmse = test(train_df, batch_size=2*batch_size)
    valid_mse, valid_rmse = test(valid_df, batch_size=2*batch_size)

    test_mse, test_rmse = test(test_df, batch_size=batch_size)
    return loss, train_mse, train_rmse, valid_mse, valid_rmse,  test_mse, test_rmse

# Example: Dynamically create learning rate schedules based on milestones
initial_lr = 10**-learning_rate
restarts = 20
decay_step = 10*5  # Decay steps for each cosine and warmup phase
warmup_factor = 0.95  # Warmup reduction factors

lr_schedule = cosineannealingwarmrestartfactor(initial_lr, restarts, decay_step, warmup_factor)

optimizer = optim.AdamW(learning_rate=lr_schedule, weight_decay=1**-weight_decay)

r = []
best_test_mse = 1e9
print(batch_size)
for e in range(epochs):
    starttime = time.time()
    
    loss, train_mse, train_rmse, valid_mse, valid_rmse, test_mse, test_rmse = epoch(e, batch_size=batch_size)
    stoptime = time.time()

    #print('RAM memory % used:', psutil.virtual_memory()[2],'RAM Used (GB):', psutil.virtual_memory()[3]/1000000000)

    best_test_mse = min(best_test_mse, test_mse)
    r.append((train_rmse,test_rmse))
    print(
        " | ".join(
            [
                f"Epoch: {e:3d}",
                f"loss: {loss:.5f}",
                f"Train mse: {train_mse:.3f}",
                f"rmse: {train_rmse:.3f}",
                f"Val mse: {valid_mse:.3f}",
                f"rmse: {valid_rmse:.3f}",
                f"Test mse: {test_mse:.3f}",
                f"rmse: {test_rmse:.3f}",
                f"LR: {np.array(optimizer.learning_rate):.6f}",
                f"Time: {stoptime-starttime}",

            ]
        )
    )
print(f"\n==> Best test mse: {best_test_mse:.3f},  rmse: {np.sqrt(best_test_mse):.3f}")
````

this produce this error now : 
````
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/4004062166.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/4004062166.py) in <cell line: 0>()
    137     starttime = time.time()
    138 
--> 139     loss, train_mse, train_rmse, valid_mse, valid_rmse, test_mse, test_rmse = epoch(e, batch_size=batch_size)
    140     stoptime = time.time()
    141 

[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/4004062166.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/4004062166.py) in epoch(e, batch_size)
    114 
    115 def epoch(e, batch_size=64):
--> 116     loss = train(train_df,e, batch_size=batch_size)
    117     train_mse, train_rmse = test(train_df, batch_size=2*batch_size)
    118     valid_mse, valid_rmse = test(valid_df, batch_size=2*batch_size)

[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/4004062166.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/4004062166.py) in train(dataset, e, batch_size, doprofile)
     72 
     73         loss = step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, y_val)
---> 74         mx.eval(state)
     75 
     76     if doprofile:

RuntimeError: [eval] Attempting to eval an array without a primitive.
If you are compiling a function, make sure all the inputs and outputs are captured:
https://ml-explore.github.io/mlx/build/html/usage/compile.html#pure-functions.
If you are not using compile, this may be a bug. Please file an issue here:
https://github.com/ml-explore/mlx/issues.

````

**Comment by thegodone** [comment]
Ok if I move the optimizer on the top I don't have the error but nan's in the loss:
`````
[50, 100, 150, 200, 250, 300, 350, 400, 450, 500, 550, 600, 650, 700, 750, 800, 850, 900, 950]
200
train
Epoch:   0 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.003113 | Time: 1.6985437870025635
train
Epoch:   1 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.002916 | Time: 0.7340600490570068
train
Epoch:   2 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.002589 | Time: 0.7343912124633789
train
Epoch:   3 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.002163 | Time: 0.7369349002838135
train
Epoch:   4 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.001680 | Time: 0.7404799461364746
train
Epoch:   5 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.001188 | Time: 0.7374305725097656
train
Epoch:   6 | loss: 0.00000 | Train mse: nan | rmse: nan | Val mse: nan | rmse: nan | Test mse: nan | rmse: nan | LR: 0.000734 | Time: 0.7343008518218994
train
---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/2359318912.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/2359318912.py) in <cell line: 0>()
    137     starttime = time.time()
    138 
--> 139     loss, train_mse, train_rmse, valid_mse, valid_rmse, test_mse, test_rmse = epoch(e, batch_size=batch_size)
    140     stoptime = time.time()
    141 

[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/2359318912.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/2359318912.py) in epoch(e, batch_size)
    123 
    124 def epoch(e, batch_size=64):
--> 125     loss = train(train_df,e, batch_size=batch_size)
    126     train_mse, train_rmse = test(train_df, batch_size=2*batch_size)
    127     valid_mse, valid_rmse = test(valid_df, batch_size=2*batch_size)

[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/2359318912.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_22122/2359318912.py) in train(dataset, e, batch_size, doprofile)
     81 
     82         loss = step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, y_val)
---> 83         mx.eval(state)
     84 
     85     if doprofile:

KeyboardInterrupt:
`````

with this code:
`````
device = mx.gpu # or mx.cpu
mx.set_default_device(device)

x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, smiles_to_rdkit_list = get_smiles_array([canonical_smiles_list[0]],feature_dicts)
num_atom_features = x_atom.shape[-1]
num_bond_features = x_bonds.shape[-1]
x_atom = mx.array(x_atom)
x_bonds = mx.array(x_bonds)
x_atom_index = mx.array(x_atom_index)
x_bond_index = mx.array(x_bond_index)
x_mask = mx.array(x_mask)

model = AttFP(radius, T, num_atom_features, num_bond_features,
            fingerprint_dim, output_units_num, p_dropout)

# Example: Dynamically create learning rate schedules based on milestones
initial_lr = 10**-learning_rate
restarts = 20
decay_step = 10*5  # Decay steps for each cosine and warmup phase
warmup_factor = 0.95  # Warmup reduction factors

lr_schedule = cosineannealingwarmrestartfactor(initial_lr, restarts, decay_step, warmup_factor)

optimizer = optim.AdamW(learning_rate=lr_schedule, weight_decay=1**-weight_decay)


from functools import partial

def loss_fn(y_hat, y):
    y = mx.reshape(y, y_hat.shape)
    return mx.mean(nn.losses.mse_loss(y_hat, y))


state = [model, optimizer.state, mx.random.state]

def forward_fn(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, labels):
    _, y_hat = model(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask)
    loss = loss_fn(y_hat, labels)
    return loss, y_hat

@partial(mx.compile, inputs=state, outputs=state)
def step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, labels):
    loss_and_grad_fn = nn.value_and_grad(model, forward_fn)
    (loss, y_hat), grads = loss_and_grad_fn(
        x_atom=x_atom,
        x_bonds=x_bonds,
        x_atom_index=x_atom_index,
        x_bond_index=x_bond_index,
        x_mask=x_mask,
        labels=labels,
    )
    optimizer.update(model, grads)
    return loss
        
def train(dataset,e, batch_size=64, doprofile=False):
    print('train')
    loss_sum = 0.0
    np.random.seed(e)
    valList = np.arange(0,dataset.shape[0])
    #shuffle them
    np.random.shuffle(valList)
    batch_list = []
    for i in range(0, dataset.shape[0], batch_size):
        batch = valList[i:i+batch_size]
        batch_list.append(batch)   
    #print('iter per batch:',len(batch_list))
    if doprofile:
        pr = cProfile.Profile()
        pr.enable()  # Start profiling

    for counter, train_batch in enumerate(batch_list):
        batch_df = dataset.loc[train_batch,:]
        smiles_list = batch_df.cano_smiles.values
        y_val = mx.array(batch_df[tasks[0]].values)
        x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, smiles_to_rdkit_list = get_smiles_array(smiles_list,feature_dicts)
        x_atom = mx.array(x_atom)
        x_bonds = mx.array(x_bonds)
        x_atom_index = mx.array(x_atom_index)
        x_bond_index = mx.array(x_bond_index)
        x_mask = mx.array(x_mask)

        loss = step(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, y_val)
        mx.eval(state)
        
    if doprofile:
        pr.disable()
        # Print profiling results
        s = io.StringIO()
        ps = pstats.Stats(pr, stream=s).sort_stats(pstats.SortKey.CUMULATIVE)
        ps.print_stats(200)  # Print top 10 results
        print(s.getvalue())    
    return loss_sum / len(dataset)

def test(test_dataset, batch_size=64):
    mse= 0.0
    valList = np.arange(0,test_dataset.shape[0])
    batch_list = []
    for i in range(0, test_dataset.shape[0], batch_size):
        batch = valList[i:i+batch_size]
        batch_list.append(batch) 
    
    model.eval()
        
    for counter, test_batch in enumerate(batch_list):
        batch_df = test_dataset.loc[test_batch,:]
        smiles_list = batch_df.cano_smiles.values
        
        x_atom, x_bonds, x_atom_index, x_bond_index, x_mask, _ = get_smiles_array(smiles_list,feature_dicts)
        x_atom = mx.array(x_atom)
        x_bonds = mx.array(x_bonds)
        x_atom_index = mx.array(x_atom_index)
        x_bond_index = mx.array(x_bond_index)
        x_mask = mx.array(x_mask)
        y_val = mx.array(batch_df[tasks[0]].values)
        _, y_hat = model(x_atom, x_bonds, x_atom_index, x_bond_index, x_mask)
        y_val = mx.reshape(y_val, y_hat.shape)
        mse += mx.square(y_hat - y_val).sum().item()
        
    val =  mse / len(test_dataset)
    model.train()
    return val, np.sqrt(val)


def epoch(e, batch_size=64):
    loss = train(train_df,e, batch_size=batch_size)
    train_mse, train_rmse = test(train_df, batch_size=2*batch_size)
    valid_mse, valid_rmse = test(valid_df, batch_size=2*batch_size)

    test_mse, test_rmse = test(test_df, batch_size=batch_size)
    return loss, train_mse, train_rmse, valid_mse, valid_rmse,  test_mse, test_rmse


r = []
best_test_mse = 1e9
print(batch_size)
for e in range(epochs):
    starttime = time.time()
    
    loss, train_mse, train_rmse, valid_mse, valid_rmse, test_mse, test_rmse = epoch(e, batch_size=batch_size)
    stoptime = time.time()

    #print('RAM memory % used:', psutil.virtual_memory()[2],'RAM Used (GB):', psutil.virtual_memory()[3]/1000000000)

    best_test_mse = min(best_test_mse, test_mse)
    r.append((train_rmse,test_rmse))
    print(
        " | ".join(
            [
                f"Epoch: {e:3d}",
                f"loss: {loss:.5f}",
                f"Train mse: {train_mse:.3f}",
                f"rmse: {train_rmse:.3f}",
                f"Val mse: {valid_mse:.3f}",
                f"rmse: {valid_rmse:.3f}",
                f"Test mse: {test_mse:.3f}",
                f"rmse: {test_rmse:.3f}",
                f"LR: {np.array(optimizer.learning_rate):.6f}",
                f"Time: {stoptime-starttime}",

            ]
        )
    )
print(f"\n==> Best test mse: {best_test_mse:.3f},  rmse: {np.sqrt(best_test_mse):.3f}")
`````

**Comment by thegodone** [comment]
On my machine M3 Max 128 GB, without our new "dropout speedup ", I have now:
Epoch:   1 | loss: 0.00000 | Train mse: 4.554 | rmse: 2.134 | Val mse: 4.166 | rmse: 2.041 | Test mse: 4.072 | rmse: 2.018 | LR: 0.002916 | Time: 0.7432119846343994

**Comment by thegodone** [comment]
One question why the loss value is always 0 now ?

**Comment by thegodone** [comment]
Now training is very fast 0.285 sec per epoch, but I have a major issue that I have seen in lot of the cases on other trial. the convergence rarely give good performance while torch and keras are much more stable. This is really a bottleneck to use MLX for me. You need to run 10 to 20 times model to get a good result while torch and keras are systematically good (0.50-0.55)

**Comment by angeloskath** [comment]
Are you using lr warmup or any scheduler? In general, it is not necessary that the hyperparameters and/or initialization will transfer 1-1 from framework to framework and you might want to play around with these a bit. If there is any reason to suspect a numerical inaccuracy then feel free to open an issue with more information on that.

**Comment by thegodone** [comment]
indeed I hope this one https://github.com/ml-explore/mlx/issues/1542 as for me the difference is significative and cannot come from just the lr warmup scheduler.

**Comment by awni** [comment]
- Did you try the suggestion in #1153 to use a linear warmup? 
- Did you try the bias corrected Adam to see if it improves the convergence rate?

If you can try either of those and report back, that would be useful to know.

> Is it possible to add a parameter for Adam optimizer to be strictly identical to pytorch / tensorflow ?

It's possible..would be good to know if it fixes your issue first though.
